### ISSUE TITLE
Hang of protobuf on OSX 'sched_yield'

### ISSUE THREAD

**Issue by wolfviking0** [google/protobuf#4203]
Hi,

I am using protobuf and I am currently facing an issue who seems to be happening only on OSX.

I have a static library with some proto class inside 'libA.a'

I have one binary application who link this libA.a and a shared library who link the same static library 'libB.dylib'.

The binary application load the dylib using dlopen RTLD_LAZY, I am loading a function using dlsym. When I call this function I have a froze/hang of my application.

The function I try to call from the shared library, initialize some of my proto file. After building in debug, I can see one of the thread are blocked here: 

0 swtch_pri
1 sched_yield
2 google::protobuf::(anonymous namespace)::SchedYield()
3 google::protobuf::(GoogleOnceInitImpl(long*, google::protobuf::Closure*)

The same exact implementation work perfectly fine on linux and I am using protobuf 3.5.1.

Do you have an idea why I have this issue on OSX ?

Thanks

**Comment by acozzette** [comment]
@wolfviking0 I often hear reports of problems with dynamic loading and I think it may be tricky to do it right. We never really load protos dynamically within Google so I don't have any experience with it, but I would guess that the problem has to do with the same code getting loaded twice. Another thing you could try would be to set `option optimize_for = LITE_RUNTIME;` in your `.proto` files and see if that makes a difference. That will disable support for reflection in the generated code and so it will simplify the initialization required.

**Comment by wolfviking0** [comment]
@acozzette I use a really dirty workaround, I commented out the SchedYield function in order to don't block.

Surprisingly, It's working and I do not have any crash. But I don't like it. I will check your suggestion but so far I think almost all my proto file are using LITE_RUNTIME.

I am thinking maybe move to FlatBuffers.

**Comment by Kixunil** [comment]
Hi, I have similarly looking problem. It only started happening lately and only one service is having this problem (we have multiple services using protobuf).

I'll investigate that library loading thing.

It seems to me that there's a problem with recursive locking during initialization. It first changes state in `once.cc` to `ONCE_STATE_EXECUTING_CLOSURE` and then it recursively calls itself and waits on it.

**Comment by Kixunil** [comment]
I've tried `option optimize_for = LITE_RUNTIME` and I got abort: `terminating with uncaught exception of type google::protobuf::FatalException: CHECK failed: generated_database_->Add(encoded_file_descriptor, size): 
abort() called`

**Comment by acozzette** [comment]
@Kixunil If `generated_database_->Add(...)` is getting called then that must mean there is some non-lite proto still being used (potentially something like `descriptor.proto` which is part of `libprotobuf`). Could you try making sure all your protos use `LITE_RUNTIME` and also make sure to link against `libprotobuf-lite` and not `libprotobuf`?

**Comment by Kixunil** [comment]
Sorry, to reply that late, I forgot to write. I've resolved it by finding out that our other team changed cmake files in such a way that it broke our binary (but not others/their binaries). I managed to get it working. I assume this is somehow related to linking, as you wrote.

Unfortunately, I don't know exactly what changed because I'm not an expert in cmake. (I just blindly edited our cmake file to match a cmake file of another team and it worked. :laughing:)

Thank you for your effort!
